File: graph/16.py
from typing import List


def is_edge(G, row, col):
    return row == 0 or row == len(G) - 1 or col == 0 or col == len(G[0]) - 1


def is_valid(G, row, col):
    return (
        row >= 0
        and row < len(G)
        and col >= 0
        and col < len(G[0]) - 1
        and G[row][col] == "O"
    )


def DFS(G, row, col):
    posRows = [-1, 0, 0, 1]
    posCols = [0, -1, 1, 0]

    G[row][col] = "#"

    for index in range(len(posRows)):
        newPosRow = row + posRows[index]
        newPosCol = col + posCols[index]

        if (
            is_valid(G, newPosRow, newPosCol)
            and is_edge(G, newPosRow, newPosCol) is False
        ):
            DFS(G, newPosRow, newPosCol)


def solve(board: List[List[str]]) -> None:
    for row in range(len(board)):
        if board[row][0] == "O":
            DFS(board, row, 0)

        if board[row][len(board[0]) - 1] == "O":
            DFS(board, row, len(board[0]) - 1)

    for col in range(len(board[0])):
        if board[0][col] == "O":
            DFS(board, 0, col)
        if board[len(board) - 1][col] == "O":
            DFS(board, len(board) - 1, col)

    for row in range(len(board)):
        for col in range(len(board[row])):
            if board[row][col] == "O":
                board[row][col] = "X"
            if board[row][col] == "#":
                board[row][col] = "O"


def print_matrix(G):
    for row in range(len(G)):
        for col in range(len(G[row])):
            print(G[row][col], end=" ")

        print()


G = [
    ["X", "X", "X", "X"],
    ["X", "O", "O", "O"],
    ["X", "X", "O", "X"],
    ["X", "O", "X", "X"],
]


print_matrix(G)
print()
solve(G)
print_matrix(G)

File: graph/16.cpp
#include <bits/stdc++.h>

using namespace std;

using Grid = vector<vector<char>>;
using Matrix = vector<vector<int>>;

vector<int> posRows = {-1, 0, 0, 1};
vector<int> posCols = {0, -1, 1, 0};

bool isEdge(const Grid &G, int row, int col)
{
    return col == 0 || col == G[0].size() - 1 || row == 0 || row == G.size() - 1;
}

bool isValid(const Grid &G, int row, int col)
{
    return row >= 0 && row < G.size() && col >= 0 && col < G[0].size() && G[row][col] == 'O';
}

void DFS(Grid &G, int row, int col)
{
    G[row][col] = '#';

    for (int i = 0; i < posRows.size(); ++i)
    {
        int newPosRow = row + posRows[i];
        int newPosCol = col + posCols[i];

        if (isValid(G, newPosRow, newPosCol) && isEdge(G, newPosRow, newPosCol) == false)
        {
            G[newPosRow][newPosCol] = 'X';
            DFS(G, newPosRow, newPosCol);
        }
    }
}

void solve(Grid &G)
{
    for (int col = 0; col < G[0].size(); ++col)
    {
        if (G[0][col] == 'O')
            DFS(G, 0, col);
        if (G[G.size() - 1][col] == 'O')
            DFS(G, G.size() - 1, col);
    }

    for (int row = 0; row < G.size(); ++row)
    {
        if (G[row][0] == 'O')
            DFS(G, row, 0);
        if (G[row][G[0].size() - 1] == 'O')
            DFS(G, row, G[0].size() - 1);
    }

    for (int i = 0; i < G.size(); ++i)
    {
        for (int j = 0; j < G[i].size(); ++j)
        {
            if(G[i][j] == 'O')
                G[i][j] = 'X';
            if(G[i][j] == '#')
                G[i][j] = 'O';
        }
    }
}

void printMatrix(const Grid &G)
{
    for (int i = 0; i < G.size(); ++i)
    {
        for (int j = 0; j < G[i].size(); ++j)
            cout << G[i][j] << " ";

        cout << "\n";
    }
}

int main()
{

    Grid G(4);
    G[0] = {'X', 'X', 'X', 'X'};
    G[1] = {'X', 'O', 'O', 'O'};
    G[2] = {'X', 'X', 'O', 'X'};
    G[3] = {'X', 'O', 'X', 'X'};

    printMatrix(G);
    cout << "\n";
    solve(G);
    printMatrix(G);

    return 0;
}